### PR TITLE
Fixes #32221 - use eslint react-hooks rules

### PR DESCRIPTION
**Pull request by jeremylenz**
Foreman was enforcing [react-hooks rules](https://reactjs.org/docs/hooks-faq.html#what-exactly-do-the-lint-rules-enforce) (rules-of-hooks and exhaustive-deps) but Katello currently is not.

This change adds those rules to Katello's eslint configuration, and makes the necessary changes to follow them.

Lint warnings generated (and fixed) here include ones like:
```
/home/vagrant/katello/webpack/components/EditableTextInput/EditableTextInput.js
  36:6  warning  React Hook useEffect has missing dependencies: 'attribute', 'inputValue', and 'onEdit'. Either include them or remove the dependency array. If 'onEdit' changes too often, find the parent component that defines it and wrap that definition in useCallback  react-hooks/exhaustive-deps

/home/vagrant/katello/webpack/components/Table/TableWrapper.js
  54:20  warning  React Hook useEffect has a spread element in its dependency array. This means we can't statically verify whether you've passed the correct dependencies  react-hooks/exhaustive-deps

/home/vagrant/katello/webpack/scenes/ContentViews/Copy/CopyContentViewForm.js
  30:7  warning  React Hook useEffect has a complex expression in the dependency array. Extract it to a separate variable so it can be statically checked  react-hooks/exhaustive-deps

/home/vagrant/katello/webpack/scenes/ContentViews/Publish/CVPublishFinish.js
  42:9  warning  The 'pollPublishTask' function makes the dependencies of useDeepCompareEffect Hook (at line 92) change on every render. Move it inside the useDeepCompareEffect callback. Alternatively, wrap the definition of 'pollPublishTask' in its own useCallback() Hook  react-hooks/exhaustive-deps

  48:9  warning  The 'handleEndTask' function makes the dependencies of useDeepCompareEffect Hook (at line 109) change on every render. To fix this, wrap the definition of 'handleEndTask' in its own useCallback() Hook                                                         react-hooks/exhaustive-deps


```

### CODE REVIEW

**Comment by jeremylenz** [comment]
For places where we use `JSON.stringify(response)` in the dependency array, we will get this:
```
 76:7  warning  React Hook useEffect has a complex expression in the dependency array. Extract it to a separate variable so it can be statically checked   react-hooks/exhaustive-deps
```
Provided that no other eslint rules are broken on those lines, I'm adding `// eslint-disable-line react-hooks/exhaustive-deps` for now.  We probably should find a better solution in the future.

**Comment by jeremylenz** [comment]
I think this will need an ack from @theforeman/packaging since we modify `package.json`

File: webpack/components/Table/TableWrapper.js
Comment (jeremylenz) [inline webpack/components/Table/TableWrapper.js:58]: still working on this one..

Comment (jeremylenz) [reply webpack/components/Table/TableWrapper.js]: Updated. All tests & lints passing now :angel:

File: package.json
Comment (tbrisker) [inline package.json:41]: could we use [eslint-plugin-foreman](https://github.com/theforeman/foreman-js/tree/master/packages/eslint-plugin-foreman) which already has this and some of the other packages, as well as some basic rules config? 
If there are rules missing there that katello uses it would be great to add there and try to unify the linting as much as possible across the whole ecosystem.

Comment (jeremylenz) [reply package.json:41]: @tbrisker This sounds fine, however I tried adding `eslint-plugin-foreman` and ran `npx eslint webpack/`.  I ran into the following:

1. It still complained about not being able to find `eslint-plugin-react-hooks`.  So I added that back in to the package.json
2. Now `eslint` complains about not being able to find module `prettier`. I see that Foreman's package.json requires `prettier`, so I also add it to Katello.  Now I get:

```
✖ 3534 problems (3534 errors, 0 warnings)
  3531 errors and 0 warnings potentially fixable with the `--fix` option.
```

Considering the huge number of changes that `prettier` will require in Katello, I think it's better to do it this way:

1. Keep `eslint-plugin-react-hooks` here
2. In a separate PR, include both the addition of `eslint-plugin-foreman` and `prettier` (cc @Andrewgdewar).   Thoughts?

Comment (jeremylenz) [reply package.json:41]: side note: I also noticed that the version of eslint-plugin-react-hooks used in `eslint-plugin-foreman` is out of date:
```
"eslint-plugin-react-hooks": "^2.1.1"
```

**Comment by jeremylenz** [comment]
Rebased after merge of https://github.com/Katello/katello/pull/9457

File: webpack/scenes/ContentViews/Details/Histories/ContentViewHistories.js
Comment (sjha4) [inline webpack/scenes/ContentViews/Details/Histories/ContentViewHistories.js:48]: Unrelated to the whole hooks rule but would it make more sense to have this in the Constants file?

File: webpack/scenes/ContentViews/Details/Versions/ContentViewVersions.js
Comment (sjha4) [inline webpack/scenes/ContentViews/Details/Versions/ContentViewVersions.js:99]: Do we also need this getContentViewVersions in the Dependency array here?

Comment (jeremylenz) [reply webpack/scenes/ContentViews/Details/Versions/ContentViewVersions.js:99]: It's defined outside of the component, so it's not needed in the dependency array according to the eslint rule.

File: webpack/scenes/ContentViews/Details/Histories/ContentViewHistories.js
Comment (jeremylenz) [reply webpack/scenes/ContentViews/Details/Histories/ContentViewHistories.js:48]: I can see that, but I feel like it makes sense to wait until it's used in multiple places.

**Comment by jeremylenz** [comment]
From IRC:

> can you highlight some of the functionality that needs regression testing? it's not totally apparrent to me just by looking at the diff

The changes here primarily affect useEffect dependency arrays, so they change whether an effect is run. Effects can contain React state updates or API calls.  If a regression were introduced, it would be causing the effect to run either more or less often than it should.  

Hopefully our React testing library tests will catch regressions.  For example, one change earlier in this PR was causing two API requests to fire every time I clicked the next page button, when there should be only one.  The test did its job and failed spectacularly with nock matching errors.

As far as manual regression testing, I can't think of any particular testing needed other than to click around the new content view pages and perform as many actions as you can.

File: package.json
Comment (Andrewgdewar) [reply package.json:41]: Sounds like a plan.

**Review by jturel — COMMENTED** [review]
This LGTM considering:

- my limited react experience
- the build is green
- the changes are made in a consistent way without any 'bonus' or confusing changes
- testing in the areas based on filename didn't expose any problems _except_ on the /hosts/register page:
```
react-dom.development.js:23093 Uncaught Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
    at checkForNestedUpdates (react-dom.development.js:23093)
    at scheduleUpdateOnFiber (react-dom.development.js:21164)
    at dispatchAction (react-dom.development.js:15660)
    at handlePluginValue (index.js:98)
    at index.js:20
    at commitHookEffectListMount (react-dom.development.js:19731)
    at commitPassiveHookEffects (react-dom.development.js:19769)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188)
    at HTMLUnknownElement.dispatchEvent (<anonymous>)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237)
```

**Comment by jturel** [comment]
Merging this will warrant a kick of green PRs containing changes in the react code to make sure they're adhering to the new rules lest we merge them and find lint errors in subsequent jenkins runs

File: webpack/scenes/ContentViews/Details/Histories/ContentViewHistories.js
Comment (Andrewgdewar) [reply webpack/scenes/ContentViews/Details/Histories/ContentViewHistories.js:48]: I feel the same way, but to be fair, I don't like that we have constants inside of our react components, I feel they should be placed above the component to prevent clutter. (I do understand the arguement of scope, but its the same memory footprint.

**Comment by jeremylenz** [comment]
Great catch @jturel!

This was a tricky one.

Turns out the `onChange` in the dependency array [here](https://github.com/Katello/katello/blob/bd611d9b4f8b38b48ffccde52d9c6ac592ba4bfe/webpack/components/extensions/RegistrationCommands/index.js#L21) was changing on every render, causing the effect to run infinitely.  So I went to apply my usual `useCallback` trick. But where is that function defined?

Turns out it's defined in Foreman thru Slot & Fill: https://github.com/theforeman/foreman/blob/71f97377b29b121dd05c84ee9993a847842e31ec/webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/index.js#L262

So our `onChange` is actually `handlePluginValue`, which is also in Foreman.

I also found that adding `handleInvalidField` [here](https://github.com/Katello/katello/blob/bd611d9b4f8b38b48ffccde52d9c6ac592ba4bfe/webpack/components/extensions/RegistrationCommands/fields/ActivationKeys.js#L43) causes a similar problem.

I've opened https://github.com/theforeman/foreman/pull/8699 which addresses both in Foreman.

For the time being, I've removed the 'problematic' dependencies and disabled the rule in those two spots here.  But when the Foreman change is merged, we can reenable it.

**Comment by sjha4** [comment]
I have played around with CV workflows on this branch and I didn't run into any issues. Changes look good to me. 👍🏽 Will kick tests on open PRs and bug you if there are things that need fixing..😄

**Comment by jturel** [comment]
> I've opened [theforeman/foreman#8699](https://github.com/theforeman/foreman/pull/8699) which addresses both in Foreman.
> 
> For the time being, I've removed the 'problematic' dependencies and disabled the rule in those two spots here. But when the Foreman change is merged, we can reenable it.

Nice fix & that totally works for me

**Review by jturel — APPROVED** [review]
ACK! Great stuff

**Pull request merged**
